tree/ntuple/v7/src/RNTupleParallelWriter.cxx

```diff
@@ -0,0 +1,156 @@
+   ColumnHandle_t AddColumn(DescriptorId_t, const RColumn &) final { return {}; }
+   void Init(RNTupleModel &) final {}
+   void UpdateSchema(const RNTupleModelChangeset &, NTupleSize_t) final
+   {
+      throw RException(R__FAIL("UpdateSchema not supported via RPageSynchronizingSink"));
+   }
+
+   void CommitPage(ColumnHandle_t, const RPage &) final
+   {
+      throw RException(R__FAIL("should never commit single pages via RPageSynchronizingSink"));
+   }
+   void CommitSealedPage(DescriptorId_t, const RSealedPage &) final
+   {
+      throw RException(R__FAIL("should never commit sealed pages via RPageSynchronizingSink"));
+   }
+   void CommitSealedPageV(std::span<RPageStorage::RSealedPageGroup> ranges) final
+   {
+      fInnerSink->CommitSealedPageV(ranges);
+   }
+   std::uint64_t CommitCluster(NTupleSize_t nNewEntries) final { return fInnerSink->CommitCluster(nNewEntries); }
+   void CommitClusterGroup() final
+   {
+      throw RException(R__FAIL("should never commit cluster group via RPageSynchronizingSink"));
+   }
+   void CommitDataset() final { throw RException(R__FAIL("should never commit dataset via RPageSynchronizingSink")); }
+
+   RPage ReservePage(ColumnHandle_t columnHandle, std::size_t nElements) final
+   {
+      return fInnerSink->ReservePage(columnHandle, nElements);
+   }
+   void ReleasePage(RPage &page) final { fInnerSink->ReleasePage(page); }
+
+   RSinkGuard GetSinkGuard() final { return RSinkGuard(fMutex); }
+};
+
+} // namespace
+
+ROOT::Experimental::RNTupleParallelWriter::RNTupleParallelWriter(std::unique_ptr<RNTupleModel> model,
+                                                                 std::unique_ptr<Detail::RPageSink> sink)
+   : fSink(std::move(sink)), fModel(std::move(model)), fMetrics("RNTupleParallelWriter")
+{
+   if (!fModel) {
+      throw RException(R__FAIL("null model"));
+   }
+   if (!fSink) {
+      throw RException(R__FAIL("null sink"));
+   }
+   fModel->Freeze();
+   fSink->Init(*fModel.get());
+   fMetrics.ObserveMetrics(fSink->GetMetrics());
+}
+
+ROOT::Experimental::RNTupleParallelWriter::~RNTupleParallelWriter()
+{
+   for (const auto &context : fFillContexts) {
+      if (!context.expired()) {
+         R__LOG_ERROR(NTupleLog()) << "RNTupleFillContext has not been destructed";
+         return;
+      }
+   }
+
+   // Now commit all clusters as a cluster group and then the dataset.
+   try {
+      fSink->CommitClusterGroup();
+      fSink->CommitDataset();
+   } catch (const RException &err) {
+      R__LOG_ERROR(NTupleLog()) << "failure committing ntuple: " << err.GetError().GetReport();
+   }
+}
+
+std::unique_ptr<ROOT::Experimental::RNTupleParallelWriter>
+ROOT::Experimental::RNTupleParallelWriter::Recreate(std::unique_ptr<RNTupleModel> model, std::string_view ntupleName,
+                                                    std::string_view storage, const RNTupleWriteOptions &options)
+{
+   auto sink = std::make_unique<Detail::RPageSinkFile>(ntupleName, storage, options);
```

Comment:
Do we want to use `RPageSink::Create()`?

Reply:
I think not: `RPageSink::Create()` might create a `RPageSinkBuf` which we don't want for the "final" sink. It's true though that it leaves out the possibility to use DAOS with the `daos://` URI scheme...

Reply:
Ah, that's right! Hm, how about an `RPagePersistentSink::Create()`? Can be a follow-up PR.

Reply:
Yes, will take a look today